libyara/modules/dex.c

```diff
@@ -0,0 +1,650 @@
+    set_integer(1, module, "invalid_dex");
+    return;
+  }
+
+  FIELD_ID_ITEM *field_ids = yr_malloc(field_ids_size);
+  memcpy(field_ids, data + offset, field_ids_size);
+
+  field_ids_size = *dex_header->field_ids_size * sizeof(FIELD_ID_ITEM);
+  int i = 0;
+  int p = 0;
+  for (i = 0, p = 0; p < field_ids_size; i += 1, p += sizeof(FIELD_ID_ITEM)) {
+    uint16_t class_idx = *field_ids[i].class_idx;
+    uint16_t type_idx = *field_ids[i].type_idx;
+    uint32_t name_idx =  *field_ids[i].name_idx;
+
+    set_integer(class_idx, module, "field_ids[%i].class_idx", i);
+    set_integer(type_idx, module, "field_ids[%i].type_idx", i);
+    set_integer(name_idx, module, "field_ids[%i].name_idx", i);
+
+    /*
+    char *class_name = get_string_item(class_idx, module);
+    char *type_name = get_string_item(get_integer(module, "type_ids[%i].descriptor_idx", type_idx), module);
+    char *field_name = get_string_item(name_idx, module);
+    printf("field(%i) = %s->%s:%s\n", i, class_name, field_name, type_name);
+    */
+  }
+  yr_free(field_ids);
+}
+
+
+void load_method_ids(PDEX_HEADER dex_header, uint8_t *data, size_t data_size, YR_OBJECT *module) {
+  uint32_t offset = *dex_header->method_ids_offset;
+  int method_ids_size = sizeof(METHOD_ID_ITEM[*dex_header->method_ids_size]);
+  if (offset + method_ids_size > data_size) {
+    set_integer(1, module, "invalid_dex");
+    return;
+  }
+
+  METHOD_ID_ITEM *method_ids = yr_malloc(method_ids_size);
+  memcpy(method_ids, data + offset, method_ids_size);
+
+  method_ids_size = *dex_header->method_ids_size * sizeof(METHOD_ID_ITEM);
+  int i, p;
+  for (i = 0, p = 0; p < method_ids_size; i += 1, p += sizeof(METHOD_ID_ITEM)) {
+    uint16_t class_idx = *method_ids[i].class_idx;
+    uint16_t proto_idx = *method_ids[i].proto_idx;
+    uint32_t name_idx =  *method_ids[i].name_idx;
+
+    set_integer(class_idx, module, "method_ids[%i].class_idx", i);
+    set_integer(proto_idx, module, "method_ids[%i].proto_idx", i);
+    set_integer(name_idx, module, "method_ids[%i].name_idx", i);
+
+    /*
+    char *class_name = get_string_item(get_integer(module, "type_ids[%i].descriptor_idx", class_idx), module);
+    char *method_name = get_string_item(name_idx, module);
+    char *prototype = get_prototype_string(proto_idx, data, module);
+    uint32_t proto_return_type_idx = get_integer(module, "proto_ids[%i].return_type_idx", proto_idx);
+    char *return_type = get_string_item(get_integer(module, "type_ids[%i].descriptor_idx", proto_return_type_idx), module);
+    printf("method(%i) = %s->%s(%s)%s\n", i, class_name, method_name, prototype, return_type);
+    */
+  }
+  yr_free(method_ids);
+}
+
+
+void load_map_list(PDEX_HEADER dex_header, uint8_t *data, size_t data_size, YR_OBJECT *module) {
+  uint32_t offset = *dex_header->map_offset;
+  uint8_t *pmap_list = data + offset;
+  size_t map_size = *pmap_list;
+  size_t map_data_size = sizeof(MAP_LIST) + (map_size * sizeof(MAP_ITEM));
```

Comment:
dont use multiply for sizes, it may overflow size_t. use calloc instead of malloc to handle this properly